interfaces/system_key.go

```diff
@@ -80,39 +119,86 @@ func generateSystemKey() *systemKey {
-func SystemKey() string {
-	sk := generateSystemKey()
+// WriteSystemKey will write the current system-key to disk
+func WriteSystemKey() error {
+	sks, err := json.Marshal(generateSystemKey())
+	if err != nil {
+		panic(err)
+	}
+	return osutil.AtomicWriteFile(dirs.SnapSystemKeyFile, sks, 0644, 0)
+}
 
-	// special case: unknown build-ids always trigger a rebuild
-	if sk.BuildID == "" {
-		return ""
+// SystemKeyMismatch checks if the running binary expects a different
+// system-key than what is on disk.
+//
+// This is used in two places:
+// - snap run: when there is a mismatch it will wait for snapd
+//             to re-generate the security profiles
+// - snapd: on startup it checks if the system-key has changed and
+//          if so re-generate the security profiles
+//
+// This ensures that "snap run" and "snapd" have a consistent set
+// of security profiles. Without it we may have the following
+// scenario:
+// 1. snapd gets refreshed and snaps need updated security profiles
+//    to work (e.g. because snap-exec needs a new permission)
+// 2. The system reboots to start the new snapd. At this point
+//    the old security profiles are on disk (because the new
+//    snapd did not run yet)
+// 3. Snaps that run as daemon get started during boot by systemd
+//    (e.g. network-manager). This may happen before snapd had a
+//    chance to refresh the security profiles.
+// 4. Because the security profiles are for the old version of
+//    the snaps that run before snapd fail to start. For e.g.
+//    network-manager this is of course catastrophic.
+// To prevent this, in step(4) we have this wait-for-snapd
+// step to ensure the expected profiles are on disk.
+func SystemKeyMismatch() (bool, error) {
+	mySystemKey := generateSystemKey()
+
+	raw, err := ioutil.ReadFile(dirs.SnapSystemKeyFile)
+	if err != nil && os.IsNotExist(err) {
+		return false, ErrSystemKeyMissing
 	}
-	sks, err := yaml.Marshal(sk)
 	if err != nil {
-		panic(err)
+		return false, err
+	}
+	var diskSystemKey systemKey
+	if err := json.Unmarshal(raw, &diskSystemKey); err != nil {
+		return false, err
+	}
+	// deal with the race that "snap run" may start, then snapd
+	// is upgraded and generates a new system-key with different
+	// inputs than the "snap run" in memory. In this case we
+	// should be fine because new security profiles will also
+	// have been written to disk.
+	if mySystemKey.Version != diskSystemKey.Version {
+		return false, ErrSystemKeyIncomparableVersions
 	}
-	return string(sks)
+
+	// special case to detect local runs
+	if mockedSystemKey == nil {
+		if exe, err := os.Readlink("/proc/self/exe"); err == nil {
+			// detect running local local builds
+			if !strings.HasPrefix(exe, "/usr") && !strings.HasPrefix(exe, "/snap") {
+				logger.Noticef("running from non-installed location %s: ignoring system-key", exe)
+				return false, ErrSystemKeyIncomparableVersions
+			}
+		}
+	}
+
+	return !reflect.DeepEqual(mySystemKey, &diskSystemKey), nil
```

Comment:
Since we have actual structs, we might compare them with code, but that's good enough as a starting point.